File: examples/js/echo_client.js
#!mojo mojo:js_content_handler
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// A JavaScript version of the C++ echo example. Connect to the echo_server.js
// application, call it's echoString() method, and then quit. To run this
// application with mojo_shell, set DIR to be the absolute path for this
// directory, then:
//   mojo_shell "file://$DIR/echo_client.js"

define("main", [
  "console",
  "mojo/services/public/js/application",
  "examples/echo/echo.mojom",
], function(console, application, echo) {

  const Application = application.Application;
  const Echo = echo.Echo;

  class EchoClientApp extends Application {
    initialize(args) {
      // The URL of this application is saved by the Application constructor.
      // Assume that the echo_server.js URL is a sibling of echo_client.js.
      // Note also: we could also connect to to the C++ echo service with:
      // var echoServerURL = "mojo:echo_server";
      var echoServerURL =
          this.url.replace("echo_client.js", "echo_server.js");
      var echoServer =
          this.shell.connectToService(echoServerURL, Echo);
      var echoClientApp = this;
      echoServer.echoString("Hello World").then(function(response) {
        console.log("Response: " + response.value);
        echoClientApp.quit();
      });
    }
  }

  return EchoClientApp;
});

File: src/examples/mybench/echo_client.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "base/logging.h"
#include "examples/echo/echo.mojom.h"
#include "mojo/public/c/system/main.h"
#include "mojo/public/cpp/application/application_impl_base.h"
#include "mojo/public/cpp/application/connect.h"
#include "mojo/public/cpp/application/run_application.h"
#include "mojo/public/cpp/utility/run_loop.h"

namespace mojo {
namespace examples {

class ResponsePrinter {
 public:
  void Run(const uint64& value) const {
    auto stop = std::chrono::high_resolution_clock::now();
    auto nanos = duration_cast<nanoseconds>(start.time_since_epoch()).count();
    std::cout << nanos - value << " ns" << std::endl;
  }
};

class EchoClientApp : public ApplicationImplBase {
 public:
  void OnInitialize() override {
    ConnectToService(shell(), "mojo:echo_server", GetProxy(&echo_));
    auto start = std::chrono::high_resolution_clock::now();
    auto nanos = std::chrono::duration_cast<std::chrono::nanoseconds>(start.time_since_epoch()).count();
    echo_->EchoString(nanos, ResponsePrinter());
  }

 private:
  EchoPtr echo_;
};

}  // namespace examples
}  // namespace mojo

MojoResult MojoMain(MojoHandle application_request) {
  mojo::examples::EchoClientApp echo_client_app;
  return mojo::RunApplication(application_request, &echo_client_app);
}
